proposals/p2022.md

```diff
@@ -0,0 +1,276 @@
+for readers.
+
+[Standard C++ Foundation's guidelines on unused parameters](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-unused)
+
+[Google's C++ style guide recommendations on unused parameters](https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions)
+
+[C++ `maybe_unused` attribute](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused)
+
+[GCC C `unused` attribute](https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html#index-g_t_0040code_007bunused_007d-attribute_002e-2640)
+
+More generally, various languages use the underscore character (`_`) to signal
+an unused binding or pattern.
+
+[Rust `_` Patterns](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#ignoring-values-in-a-pattern)
+
+[Golang Blank Identifiers](https://go.dev/ref/spec#Blank_identifier)
+
+[Scala Wildcard Patterns](https://www.scala-lang.org/files/archive/spec/2.11/10-pattern-matching.html#variable-patterns)
+
+[Crystal `_` Cases](https://crystal-lang.org/reference/1.5/syntax_and_semantics/case.html#underscore-allowed)
+
+## Proposal
+
+Users can explicitly declare a value in a pattern as unused in two forms:
+
+-   `_: Int` : Using the underscore token, `_`, in place of a name.
+
+-   `unused size: Int` : Using the leading `unused` keyword followed by the
+    name.
+
+## Details
+
+Introducing two syntaxes satisfies the desire to have a terse way to discard
+values, but still provides authors with a more verbose, explicit syntax that
+preserves the name.
+
+Both approaches are unambiguous -- to human readers and authors as well as
+programmatic interpretations. The inclusion of an explicit `unused` keyword
+allows authors to preserve the name of a value for documentation purposes, while
+still explicitly marking the value as discarded in an interpretable way to
+humans and programs alike.
+
+The `unused` keyword would be a new keyword in Carbon. This keyword would only
+be valid when preceding a name in a pattern binding and the keyword would
+tightly bind to the following name, disallowing specifying an entire sub-pattern
+as `unused`.
+
+### The behavior of `unused` name bindings
+
+Names that are qualified with the `unused` keyword are visible for name lookup
+but uses are invalid, including when they cause ambiguous name lookup errors. If
+attempted to be used, a compiler error will be shown to the user, instructing
+them to either remove the `unused` qualifier or remove the use.
+
+The inverse, where a name is not qualified by the `unused` qualifier, but never
+used, will cause the compiler to emit a warning diagnostic, informing the user
+that a given name was not used and suggesting to either remove the binding or
+mark it as unused.
+
+### Examples
+
+Discarded Name : Function Parameter
+
+```carbon
+// Interface
+fn Sum(x: List(Int), size: Int) -> Int;
+
+// Implementation that doesn't use the size parameter
+fn Sum(x: List(Int), _: Int) -> Int;
```

Comment:
```suggestion
fn Sum(x: List(Int), _: Int) -> Int { ... }
// or:
fn Sum(x: List(Int), unused size: Int) -> Int { ... }
```
This "or" is instead of the other "Function Parameter" example below.